sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/FlatMapElements.java

```diff
@@ -170,4 +174,193 @@ public void populateDisplayData(DisplayData.Builder builder) {
+     *         .via(ee -> e.exception().getMessage()));
+     * PCollection<String> errors = result.errors();
+     * }</pre>
+     */
+    public FlatMapWithExceptions<InputT, OutputT, FailureT> via(
+        ProcessFunction<ExceptionElement<InputT>, FailureT> exceptionHandler) {
+      return new FlatMapWithExceptions<>(
+          fn, originalFnForDisplayData, inputType, outputType, exceptionHandler, failureType);
+    }
+
+    /**
+     * Like {@link #via(ProcessFunction)}, but takes advantage of the type information provided by
+     * {@link InferableFunction}, meaning that a call to {@link #into(TypeDescriptor)} may not be
+     * necessary.
+     *
+     * <p>Example usage:
+     *
+     * <pre>{@code
+     * Result<PCollection<Integer>, KV<String, Map<String, String>>> result = words.apply(
+     *     FlatMapElements.into(TypeDescriptors.strings())
+     *         .via((String line) -> Arrays.asList(Arrays.copyOfRange(line.split(" "), 1, 5)))
+     *         .withExceptions()
+     *         .via(new WithExceptions.ExceptionAsMapHandler<String>() {}));
+     * PCollection<KV<String, Map<String, String>>> errors = result.errors();
+     * }</pre>
+     */
+    public <NewFailureT> FlatMapWithExceptions<InputT, OutputT, NewFailureT> via(
+        InferableFunction<ExceptionElement<InputT>, NewFailureT> exceptionHandler) {
+      return new FlatMapWithExceptions<>(
+          fn,
+          originalFnForDisplayData,
+          inputType,
+          outputType,
+          exceptionHandler,
+          exceptionHandler.getOutputTypeDescriptor());
+    }
+
+    @Override
+    public WithExceptions.Result<PCollection<OutputT>, FailureT> expand(PCollection<InputT> input) {
+      final TupleTag<OutputT> outputTag = new TupleTag<OutputT>() {};
+      final TupleTag<FailureT> failureTag;
+      if (failureType == null) {
+        failureTag = new TupleTag<>();
+      } else {
+        failureTag =
+            new TupleTag<FailureT>() {
+              @Override
+              public TypeDescriptor<FailureT> getTypeDescriptor() {
+                return failureType;
+              }
+            };
+      }
+      DoFn<InputT, OutputT> doFn =
+          new DoFn<InputT, OutputT>() {
+            @ProcessElement
+            public void processElement(
+                @Element InputT element, MultiOutputReceiver receiver, ProcessContext c)
+                throws Exception {
+              boolean exceptionWasThrown = false;
+              Iterable<OutputT> res = null;
+              try {
+                res = fn.getClosure().apply(c.element(), Fn.Context.wrapProcessContext(c));
+              } catch (Exception e) {
+                exceptionWasThrown = true;
+                ExceptionElement<InputT> exceptionElement = ExceptionElement.of(element, e);
+                receiver.get(failureTag).output(exceptionHandler.apply(exceptionElement));
+              }
+              if (!exceptionWasThrown) {
+                for (OutputT output : res) {
```

Comment:
I'm hesitant to put anything into the `try` except for the single line where we apply `fn`. I have similar code to this running in a Dataflow pipeline, and I originally sent elements to the main output receiver within the `try` block, but I saw instances where some elements were appearing in both the output and error collections. When I factored the receiver code out of the try block, I saw pipeline exceptions bubbling all the way up.

I don't have a full understanding of what was happening, but I think it may be related to fusion; essentially, I think the try block was catching an exception raised by code from an adjacent transform. I don't know anything about how fusion is implemented, but is it possible that calling `.output` on a receiver is actually calling code from a fused transform?

Reply:
Ah, this is a very good point. I believe that the output() will immediately call the next fused DoFn with that element, so errors will bubble all the way up.